File: leet_lint_code/leet_code/four_sum.cpp
// Copyright © 2018 Hailin Gu. All rights reserved.
// License(GPL)
// Author: Hailin Gu
// This is a answer of leet code problem 18.

#ifndef LEETLINTCODE_LINTCODE_FOURSUM_CPP
#define LEETLINTCODE_LINTCODE_FOURSUM_CPP

#include <vector>
#include <algorithm>

class FourSum {
 public:
  std::vector<std::vector<int>> fourSum(std::vector<int> num, int target) {
    std::vector<std::vector<int> > ret;
    std::sort(num.begin(),num.end());
    int len = (int)num.size();
    for (int i = 0; i < len - 3; ++ i) {
      if (num[i] + num[len - 1] + num[len - 2] + num[len-3] < target) continue;
      std::vector<std::vector<int> > tmp = threeSum(num, target - num[i], num[i], i + 1);
      for (int j = 0; j < tmp.size(); ++ j) ret.push_back(tmp[j]);
      while(num[i] == num[i + 1]) ++ i;
    }
    return ret;
  }

  inline std::vector<std::vector<int> > threeSum(std::vector<int> &num, int target, int &val, int index) {
    std::vector<std::vector<int> > ret;
    std::sort(num.begin(),num.end());
    int len = (int)num.size(), j, k;
    for(int i = index; i < len - 2; ++ i){
      j = i + 1, k = len - 1;
      while (j < k) {
        int sum = num[i] + num[j] + num[k];
        if(sum > target){
          while (num[k] == num[k - 1]) { -- k; }
          -- k;
        }else if(sum < target){
          while (num[j] == num[j + 1]) { ++ j;}
          ++ j;
        }else{
          ret.push_back({val, num[i], num[j], num[k]});
          while (num[j] == num[j + 1]) {++j;}
          ++ j;
        }
      }
      while (num[i] == num[i + 1]) ++i;
    }
    return ret;
  }
};

#endif // LEETLINTCODE_LINTCODE_FOURSUM_CPP

File: leet_lint_code/leet_code/java/FourSum.java
// Copyright © 2018 Hailin Gu. All rights reserved.
// License(GPL)
// Author: Hailin Gu
// This is a answer of leet code problem 18.
// Date: 2019.8.26

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class FourSum {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        List<List<Integer>> ans = new ArrayList<>();
        Arrays.sort(nums);

        for (int i = 0; i < nums.length - 3; ) {
            for (int j = i + 1; j < nums.length - 2; ) {
                int left = j + 1;
                int right = nums.length - 1;

                while (left < right) {
                    int sum = nums[i] + nums[j] + nums[left] + nums[right];
                    if (sum == target) {
                        List<Integer> list = new ArrayList<>();
                        list.add(nums[i]);
                        list.add(nums[j]);
                        list.add(nums[left]);
                        list.add(nums[right]);
                        ans.add(list);
                        while (right >= 1 && nums[right - 1] == nums[right]) -- right;
                        -- right;
                        while (left < nums.length - 1 && nums[left] == nums[left + 1]) ++ left;
                        ++ left;
                    } else if (sum < target) {
                        while (left < nums.length - 1 && nums[left] == nums[left + 1]) ++ left;
                        ++ left;
                    } else {
                        while (right >= 1 && nums[right - 1] == nums[right]) -- right;
                        -- right;
                    }
                }
                while (j < nums.length - 1 && nums[j + 1] == nums[j]) ++ j;
                ++ j;
            }
            while (i < nums.length - 1 && nums[i + 1] == nums[i]) ++ i;
            ++ i;
        }
        return ans;
    }
}
